app/ui/browser/actions/main-actions.js

```diff
@@ -77,18 +99,54 @@ export function unbookmark(url) {
   };
 }
 
-export function navigatePageTo(pageId, location) {
-  return { type: types.NAVIGATE_PAGE_TO, pageId, location };
+// @TODO Linting https://github.com/eslint/eslint/issues/3587
+export function navigatePageTo(pageId, location, doc=document) { // eslint-disable-line
+  return tagErrors((dispatch, getState) => {
+    assert(typeof location === 'string', '`location` must be a string.');
+    assert(isUUID(pageId), `Invalid page id: ${pageId}`);
+    assert(getPageIndexById(getState(), pageId) >= 0,
+      `Page ${pageId} not found in current state.`);
+
+    dispatch({ type: types.NAVIGATE_PAGE_TO, pageId, location });
+    getWebViewById(doc, pageId).setAttribute('src', fixURL(location));
+  }, types.NAVIGATE_PAGE_TO);
 }
 
-export function navigatePageBack(pageId) {
-  return { type: types.NAVIGATE_PAGE_BACK, pageId };
+// @TODO Linting https://github.com/eslint/eslint/issues/3587
+export function navigatePageBack(pageId, doc=document) { // eslint-disable-line
+  return tagErrors((dispatch, getState) => {
+    assert(isUUID(pageId), `Invalid page id: ${pageId}`);
+    const pageIndex = getPageIndexById(getState(), pageId);
+    assert(pageIndex >= 0, `Page ${pageId} not found in current state.`);
+    assert(getState().browserWindow.pages.get(pageIndex).canGoBack, 'Page cannot go back.');
+
+    dispatch({ type: types.NAVIGATE_PAGE_BACK, pageId });
+    getWebViewById(doc, pageId).goBack();
+  }, types.NAVIGATE_PAGE_BACK);
 }
 
-export function navigatePageForward(pageId) {
-  return { type: types.NAVIGATE_PAGE_FORWARD, pageId };
+// @TODO Linting https://github.com/eslint/eslint/issues/3587
+export function navigatePageForward(pageId, doc=document) { // eslint-disable-line
+  return tagErrors((dispatch, getState) => {
+    assert(isUUID(pageId), `Invalid page id: ${pageId}`);
+    const pageIndex = getPageIndexById(getState(), pageId);
+    assert(pageIndex >= 0, `Page ${pageId} not found in current state.`);
+    assert(getState().browserWindow.pages.get(pageIndex).canGoForward, 'Page cannot go forward.');
+
+    dispatch({ type: types.NAVIGATE_PAGE_FORWARD, pageId });
+    getWebViewById(doc, pageId).goForward();
+  }, types.NAVIGATE_PAGE_FORWARD);
 }
 
-export function navigatePageRefresh(pageId) {
-  return { type: types.NAVIGATE_PAGE_REFRESH, pageId };
+// @TODO Linting https://github.com/eslint/eslint/issues/3587
+export function navigatePageRefresh(pageId, doc=document) { // eslint-disable-line
+  return tagErrors((dispatch, getState) => {
+    assert(isUUID(pageId), `Invalid page id: ${pageId}`);
+    const pageIndex = getPageIndexById(getState(), pageId);
+    assert(pageIndex >= 0, `Page ${pageId} not found in current state.`);
+    assert(getState().browserWindow.pages.get(pageIndex).canRefresh, 'Page cannot refresh.');
+
+    dispatch({ type: types.NAVIGATE_PAGE_REFRESH, pageId });
```

Comment:
Redux, and a lot of the ecosystem around Redux, assume that the reducer has no side effects. From the Redux docs:

> It’s very important that the reducer stays pure. Things you should never do inside a reducer:
>
>    Mutate its arguments;
>    Perform side effects like API calls and routing transitions;
>    Call non-pure functions, e.g. Date.now() or Math.random().

I don't think we'll be able to achieve replayability for these actions. I might be seeing it wrong, so if you think we can maybe we should walk through a use case.